### PR TITLE
Hotfix/pms fail count restart

### PR DESCRIPTION
**Pull request by pnt325**
Restart device after device read PMS5003 sensor failed 10 times. #217

### CODE REVIEW

File: examples/BASIC/BASIC.ino [hunk 1]
```diff
@@ -514,16 +513,21 @@ static void updatePm(void) {
...
     Serial.printf("PM10 ug/m3: %d\r\n", measurements.pm10_1);
     Serial.printf("PM0.3 Count: %d\r\n", measurements.pm03PCount_1);
-    pmFailCount = 0;
+    ag.pms5003.resetFailCount();
   } else {
-    pmFailCount++;
-    Serial.printf("PMS read failed: %d\r\n", pmFailCount);
-    if (pmFailCount >= 3) {
+    ag.pms5003.updateFailCount();
+    Serial.printf("PMS read failed: %d\r\n", ag.pms5003.getFailCount());
```
Comment (nick-4711) [inline examples/BASIC/BASIC.ino:519]: Propose to rephrase to

PMS read failed %d times

File: examples/BASIC/BASIC.ino [hunk 2]
```diff
@@ -514,16 +513,21 @@ static void updatePm(void) {
...
+    if (ag.pms5003.getFailCount() >= PMS_FAIL_COUNT_SET_INVALID) {
       measurements.pm01_1 = utils::getInvalidPMS();
```
Comment (nick-4711) [inline examples/BASIC/BASIC.ino:521]: The method name is misleading propose  to write
getInvalidPmValue(), maybe we can also use a constant

File: examples/BASIC/BASIC.ino [hunk 3]
```diff
@@ -514,16 +513,21 @@ static void updatePm(void) {
...
       measurements.pm25_1 = utils::getInvalidPMS();
       measurements.pm10_1 = utils::getInvalidPMS();
       measurements.pm03PCount_1 = utils::getInvalidPMS();
     }
+
+    if(ag.pms5003.getFailCount() >= ag.pms5003.getFailCountMax()) {
```
Comment (nick-4711) [inline examples/BASIC/BASIC.ino:527]: code formatting issue

File: examples/BASIC/BASIC.ino [hunk 4]
```diff
@@ -514,16 +513,21 @@ static void updatePm(void) {
...
+      Serial.printf("Restarting...");
```
Comment (nick-4711) [inline examples/BASIC/BASIC.ino:528]: Propose to write why we restart, e.g. something like
PMS failure count is %d, the max is set to %d, restarting ....

File: examples/OneOpenAir/OneOpenAir.ino [hunk 5]
```diff
@@ -1053,13 +1057,26 @@ static void updatePm(void) {
...
     } else {
-      measurements.pm01_1 = utils::getInvalidPMS();
-      measurements.pm25_1 = utils::getInvalidPMS();
-      measurements.pm10_1 = utils::getInvalidPMS();
-      measurements.pm03PCount_1 = utils::getInvalidPMS();
-      measurements.temp_1 = utils::getInvalidTemperature();
-      measurements.hum_1 = utils::getInvalidHumidity();
+      if (configuration.hasSensorPMS1) {
+        ag->pms5003t_1.updateFailCount();
+        Serial.printf("[1] PMS read failed: %d\r\n", ag->pms5003t_1.getFailCount());
```
Comment (nick-4711) [inline examples/OneOpenAir/OneOpenAir.ino:1065]: The message is misleading, the read hasn't failed here?

File: examples/OneOpenAir/OneOpenAir.ino [hunk 6]
```diff
@@ -1006,6 +1005,7 @@ static void updateTvoc(void) {
 }
 
 static void updatePm(void) {
+  bool failToRestart = false;
```
Comment (nick-4711) [inline examples/OneOpenAir/OneOpenAir.ino:1008]: propose to rename to simply "restart"

File: src/PMS/PMS.cpp [hunk 7]
```diff
@@ -147,6 +148,36 @@ void PMSBase::handle() {
...
+  if (failCount < failCountMax) {
+    failCount++;
+  }
+}
+
+/**
+ * @brief Reset number of fail
+ * 
+ */
+void PMSBase::resetFailCount(void) {failCount = 0;}
```
Comment (nick-4711) [inline src/PMS/PMS.cpp:165]: Code format issue

File: src/PMS/PMS.cpp [hunk 8]
```diff
@@ -147,6 +148,36 @@ void PMSBase::handle() {
  */
 bool PMSBase::isFailed(void) { return failed; }
 
+/**
+ * @brief Increate number of fail
```
Comment (nick-4711) [inline src/PMS/PMS.cpp:152]: Increase vs. update. However, we can remove such comments they don't provide any value.

File: src/PMS/PMS.cpp [hunk 9]
```diff
@@ -147,6 +148,36 @@ void PMSBase::handle() {
...
+}
+
+/**
+ * @brief Reset number of fail
+ * 
+ */
+void PMSBase::resetFailCount(void) {failCount = 0;}
+
+/**
+ * @brief Get number of fail
```
Comment (nick-4711) [inline src/PMS/PMS.cpp:168]: remove

File: src/PMS/PMS.cpp [hunk 10]
```diff
@@ -147,6 +148,36 @@ void PMSBase::handle() {
...
+ * 
+ * @return int 
+ */
+int PMSBase::getFailCount(void) { return failCount; }
+
+/**
+ * @brief Get maximum of fail count max
```
Comment (nick-4711) [inline src/PMS/PMS.cpp:175]: remove

File: src/PMS/PMS5003.cpp [hunk 11]
```diff
@@ -175,3 +175,33 @@ void PMS5003::handle(void) { pms.handle(); }
  * @return false Communication timeout or sensor has removed
  */
 bool PMS5003::isFailed(void) { return pms.isFailed(); }
+
+/**
+ * @brief Increate number of fail
```
Comment (nick-4711) [inline src/PMS/PMS5003.cpp:180]: remove

File: src/PMS/PMS5003.cpp [hunk 12]
```diff
@@ -175,3 +175,33 @@ void PMS5003::handle(void) { pms.handle(); }
...
+ * 
+ */
+void PMS5003::updateFailCount(void) {
+  pms.updateFailCount();
+}
+
+/**
+ * @brief Reset fail count
```
Comment (nick-4711) [inline src/PMS/PMS5003.cpp:188]: remove

File: examples/OneOpenAir/OneOpenAir.ino (hunk 5 above)
Comment (pnt325) [reply examples/OneOpenAir/OneOpenAir.ino:1065]: it's failed if `configuration.hasSensorPMS1`=`true`

**Review by nick-4711 — APPROVED** [review]

**Pull request merged**